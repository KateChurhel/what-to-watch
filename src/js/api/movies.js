/* istanbul ignore file */
import { sendGetRequest } from './index';

export const getCategoryData = (category, data) => sendGetRequest(`/${category}/popular`, data);

export const getDetailsData = (category, id, data) => sendGetRequest(`/${category}/${id}`, data);

export const getSearchResult = (category, data) => sendGetRequest(`/search/${category}`, data);

export const getDiscoverResult = (data) => sendGetRequest('/discover/movie', data);

export const getGenres = () => sendGetRequest('/genre/movie/list');
