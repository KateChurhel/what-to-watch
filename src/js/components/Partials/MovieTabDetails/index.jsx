// libraries
import React from 'react';
import PropTypes from 'prop-types';
import { useParams } from 'react-router-dom';
// constants
import CATEGORIES from '../../../constants/categories';

const MovieTabDetails = ({
  numberOfEpisodes, numberOfSeasons, tagline, homepage, cardTitle, originalTitle,
  originalLanguage, formattedRunTime, genresList, releaseDate, status,
}) => {
  const { category } = useParams();
  const isMovie = category === CATEGORIES.movie;

  return (
    <>
      <p className="movie-card__details-item">
        <strong>Title: </strong>
        {cardTitle}
      </p>
      <p className="movie-card__details-item">
        <strong>Original Title: </strong>
        {originalTitle}
      </p>
      {tagline && (
      <p className="movie-card__details-item">
        <strong>Quote: </strong>
        {tagline}
      </p>
      )}
      {!isMovie && (
        <>
          <p className="movie-card__details-item">
            <strong>Seasons: </strong>
            {numberOfSeasons}
          </p>
          <p className="movie-card__details-item">
            <strong>Episodes: </strong>
            {numberOfEpisodes}
          </p>
        </>
      )}
      <p className="movie-card__details-item">
        <strong>
          {`${isMovie ? '' : 'Episode '} Run Time `}
        </strong>
        {formattedRunTime}
      </p>
      <p className="movie-card__details-item">
        <strong>Genre: </strong>
        {genresList}
      </p>
      <p className="movie-card__details-item">
        <strong>Language: </strong>
        {originalLanguage}
      </p>
      <p className="movie-card__details-item">
        <strong>Status: </strong>
        {status}
      </p>
      {releaseDate && (
      <p className="movie-card__details-item">
        <strong>Release date: </strong>
        {releaseDate}
      </p>
      )}
      {homepage ? (
        <p className="movie-card__details-item">
          <strong>Homepage: </strong>
          <a href={homepage} rel="noreferrer noopener" target="_blank">Link</a>
        </p>
      ) : null}
    </>
  );
};

MovieTabDetails.defaultProps = {
  numberOfEpisodes: 0,
  numberOfSeasons: 0,
  tagline: '',
  homepage: '',
  cardTitle: '',
  originalTitle: '',
  originalLanguage: '',
  formattedRunTime: '',
  genresList: '',
  releaseDate: '',
  status: '',
};

MovieTabDetails.propTypes = {
  numberOfEpisodes: PropTypes.number,
  numberOfSeasons: PropTypes.number,
  tagline: PropTypes.string,
  homepage: PropTypes.string,
  cardTitle: PropTypes.string,
  originalTitle: PropTypes.string,
  originalLanguage: PropTypes.string,
  formattedRunTime: PropTypes.string,
  genresList: PropTypes.string,
  releaseDate: PropTypes.string,
  status: PropTypes.string,
};

export default MovieTabDetails;
