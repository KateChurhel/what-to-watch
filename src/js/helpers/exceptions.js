// constants
import DEFAULT_ERROR from '../constants/errors';

export class RequestError extends Error {
  constructor(status = '', responseData) {
    super();

    const errorData = responseData || {};

    this.message = errorData.message || DEFAULT_ERROR;
    this.errors = errorData.errors;
    this.status = status;
    this.name = 'RequestError';
  }
}

export default {};
