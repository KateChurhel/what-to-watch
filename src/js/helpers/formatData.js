import moment from 'moment';
import 'moment-duration-format';

export const formatTime = (time, unit, format) => moment.duration(time, unit).format(format, { trim: false });
export const formatDate = (date, format) => moment(date).format(format);

export const parseQueryString = (queryString) => {
  const query = {};
  const pairs = (queryString[0] === '?' ? queryString.substr(1) : queryString).split('&');
  for (let i = 0; i < pairs.length; i++) {
    const pair = pairs[i].split('=');
    query[decodeURIComponent(pair[0])] = decodeURIComponent(pair[1] || '');
  }

  return query;
};

export const getStringValues = (array = [], key) => (array.length ? array.map((item) => item[key]).join(', ') : '');

export const formatCardData = ({
  number_of_episodes: numberOfEpisodes, number_of_seasons: numberOfSeasons, original_name: originalName,
  episode_run_time: episodeRunTime, reviews: { results: reviews = [] } = {}, homepage, original_title: originalTitle,
  original_language: originalLanguage, spoken_languages: spokenLanguages, status, runtime, tagline,
  release_date: releaseDate, genres, poster_path: poster, backdrop_path: backgroundImage,
  vote_average: voteAverage, vote_count: voteCount, overview, title, name, id,
  credits: { cast = [] } = {}, recommendations: { results: recommendations = [] } = {},
} = {}) => {
  const formattedRunTime = formatTime(runtime || episodeRunTime, 'minutes', 'h[h] m[m]');
  const releaseYear = formatDate(releaseDate, 'YYYY');
  const genresList = getStringValues(genres, 'name');
  const spokenLanguagesList = getStringValues(spokenLanguages, 'name');

  return {
    id,
    numberOfEpisodes,
    numberOfSeasons,
    episodeRunTime,
    reviews,
    homepage,
    originalLanguage,
    status,
    tagline,
    releaseDate,
    poster,
    backgroundImage,
    voteAverage,
    voteCount,
    overview,
    cast,
    recommendations,
    originalTitle: originalTitle || originalName,
    cardTitle: title || name,
    formattedRunTime,
    releaseYear,
    genresList,
    spokenLanguagesList,
  };
};
