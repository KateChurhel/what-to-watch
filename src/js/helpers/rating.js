import USER_RATINGS from '../constants/rating';

export const computeRatingLevel = (rating) => USER_RATINGS.filter(
  (rank) => rank.min <= rating,
).pop().title;
