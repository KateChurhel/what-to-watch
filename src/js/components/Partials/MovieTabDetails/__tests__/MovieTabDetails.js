import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter, Route } from 'react-router-dom';
import CATEGORIES from '../../../../constants/categories';
import ROUTES from '../../../../constants/routes';
import { MOCK_MOVIES, MOCK_TVS } from '../../../../mocks/movies';
import MovieTabDetails from '../index';
import { formatCardData } from '../../../../helpers/formatData';

describe('Movies Tab Details: ', () => {
  it('correctly renders with movie data', async () => {
    const { asFragment } = render(
      <MemoryRouter initialEntries={[`/${CATEGORIES.movie}/1`]} initialIndex={1}>
        <Route path={ROUTES.catalog}>
          <MovieTabDetails {...formatCardData(MOCK_MOVIES[0])} />
        </Route>
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('correctly renders with tv data', async () => {
    const { asFragment } = render(
      <MemoryRouter initialEntries={[`/${CATEGORIES.tv}/1`]} initialIndex={1}>
        <Route path={ROUTES.catalog}>
          <MovieTabDetails {...formatCardData(MOCK_TVS[0])} />
        </Route>
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('correctly renders without data', async () => {
    const { asFragment } = render(
      <MemoryRouter initialEntries={[`/${CATEGORIES.movie}/1`]} initialIndex={1}>
        <Route path={ROUTES.catalog}>
          <MovieTabDetails />
        </Route>
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
