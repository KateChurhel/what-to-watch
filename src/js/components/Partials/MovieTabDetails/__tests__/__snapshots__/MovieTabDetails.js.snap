// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movies Tab Details:  correctly renders with movie data 1`] = `
<DocumentFragment>
  <p
    class="movie-card__details-item"
  >
    <strong>
      Title: 
    </strong>
    La La Land
  </p>
  <p
    class="movie-card__details-item"
  >
    <strong>
      Original Title: 
    </strong>
    La La Land
  </p>
  <p
    class="movie-card__details-item"
  >
    <strong>
      Quote: 
    </strong>
    Here's to the fools who dream.
  </p>
  <p
    class="movie-card__details-item"
  >
    <strong>
       Run Time 
    </strong>
    2h 9m
  </p>
  <p
    class="movie-card__details-item"
  >
    <strong>
      Genre: 
    </strong>
    Comedy, Drama, Romance, Music
  </p>
  <p
    class="movie-card__details-item"
  >
    <strong>
      Language: 
    </strong>
    en
  </p>
  <p
    class="movie-card__details-item"
  >
    <strong>
      Status: 
    </strong>
    Released
  </p>
  <p
    class="movie-card__details-item"
  >
    <strong>
      Release date: 
    </strong>
    2016-11-29
  </p>
  <p
    class="movie-card__details-item"
  >
    <strong>
      Homepage: 
    </strong>
    <a
      href="http://www.lalaland.movie"
      rel="noreferrer noopener"
      target="_blank"
    >
      Link
    </a>
  </p>
</DocumentFragment>
`;

exports[`Movies Tab Details:  correctly renders with tv data 1`] = `
<DocumentFragment>
  <p
    class="movie-card__details-item"
  >
    <strong>
      Title: 
    </strong>
    Supernatural
  </p>
  <p
    class="movie-card__details-item"
  >
    <strong>
      Original Title: 
    </strong>
    Supernatural
  </p>
  <p
    class="movie-card__details-item"
  >
    <strong>
      Seasons: 
    </strong>
    15
  </p>
  <p
    class="movie-card__details-item"
  >
    <strong>
      Episodes: 
    </strong>
    327
  </p>
  <p
    class="movie-card__details-item"
  >
    <strong>
      Episode  Run Time 
    </strong>
    0h 45m
  </p>
  <p
    class="movie-card__details-item"
  >
    <strong>
      Genre: 
    </strong>
    Drama, Mystery, Sci-Fi & Fantasy
  </p>
  <p
    class="movie-card__details-item"
  >
    <strong>
      Language: 
    </strong>
    en
  </p>
  <p
    class="movie-card__details-item"
  >
    <strong>
      Status: 
    </strong>
    Returning Series
  </p>
  <p
    class="movie-card__details-item"
  >
    <strong>
      Homepage: 
    </strong>
    <a
      href="http://www.cwtv.com/shows/supernatural"
      rel="noreferrer noopener"
      target="_blank"
    >
      Link
    </a>
  </p>
</DocumentFragment>
`;

exports[`Movies Tab Details:  correctly renders without data 1`] = `
<DocumentFragment>
  <p
    class="movie-card__details-item"
  >
    <strong>
      Title: 
    </strong>
  </p>
  <p
    class="movie-card__details-item"
  >
    <strong>
      Original Title: 
    </strong>
  </p>
  <p
    class="movie-card__details-item"
  >
    <strong>
       Run Time 
    </strong>
  </p>
  <p
    class="movie-card__details-item"
  >
    <strong>
      Genre: 
    </strong>
  </p>
  <p
    class="movie-card__details-item"
  >
    <strong>
      Language: 
    </strong>
  </p>
  <p
    class="movie-card__details-item"
  >
    <strong>
      Status: 
    </strong>
  </p>
</DocumentFragment>
`;
